File: 11_动态规划/经典题/LCP 69. Hello LeetCode!.py
"""https://leetcode.cn/problems/rMeRt2/

# 参观者只需要集齐 helloleetcode 的 13 张字母卡片即可获得力扣纪念章。
# 在展位上有一些由字母卡片拼成的单词,words[i][j] 表示第 i 个单词的第 j 个字母。
# 你可以从这些单词中取出一些卡片,但每次拿取卡片都需要消耗游戏代币,规则如下：

# !从一个单词中取一个字母所需要的代币数量,为该字母左边和右边字母数量之积
# !可以从一个单词中多次取字母,每个字母仅可被取一次
# 例如：从 example 中取出字母 a,需要消耗代币 2*4=8,字母取出后单词变为 exmple:
# 再从中取出字母 m,需要消耗代币 2*3=6,字母取出后单词变为 exple:
# !请返回取得 helloleetcode 这些字母需要消耗代币的 最少 数量。如果无法取得,返回 -1。
# 取出字母的顺序没有要求
# !取出的所有字母`恰好`可以拼成 helloleetcode

# 1 <= words.length <= 24
# 1 <= words[i].length <= 8

总结:
# !1.dp[index][c][d][e][h][l][o][t] 表示从 index 开始,取出的各个字母的数量,一共24*960个状态
# !2.每次转移枚举需要删除哪几个位置的字母,然后计算代价,取最小值
  注意这个代价是可以预处理的、枚举子集也是可以预处理的
  转移复杂度O(2^8*8)
"""

from functools import lru_cache
from collections import deque
from typing import List, Tuple


INF = int(1e12)


@lru_cache(None)
def calCost(need: Tuple[bool, ...]) -> int:
    """计算从两端取出所需字符的代价 双指针"""
    n = len(need)
    left, right = 0, n - 1
    cost = 0
    leftMoved, rightMoved = 0, 0
    while left <= right:
        while left <= right and not need[left]:
            left += 1
        while left <= right and not need[right]:
            right -= 1
        if left > right:
            break
        leftCost = (left - leftMoved) * (n - 1 - left - rightMoved)
        rightCost = (right - leftMoved) * (n - 1 - right - rightMoved)
        if leftCost <= rightCost:
            left += 1
            cost += leftCost
            leftMoved += 1
        else:
            right -= 1
            cost += rightCost
            rightMoved += 1
    return cost


# @lru_cache(None)
# def calCost(need: Tuple[bool, ...]) -> int:
#     """计算从两端取出所需字符的代价 双指针"""
#     n = len(need)
#     left, right = 0, n - 1
#     cost = 0
#     leftMoved, rightMoved = 0, 0
#     while left <= right:
#         while left <= right and not need[left]:
#             left += 1
#         while left <= right and not need[right]:
#             right -= 1
#         if left > right:
#             break
#         leftCost = (left - leftMoved) * (n - 1 - left - rightMoved)
#         rightCost = (right - leftMoved) * (n - 1 - right - rightMoved)
#         if leftCost <= rightCost:
#             left += 1
#             cost += leftCost
#             leftMoved += 1
#         else:
#             right -= 1
#             cost += rightCost
#             rightMoved += 1
#     return cost


CDETOLH = set("cdetolh")
ORDER = {
    "c": 0,
    "d": 1,
    "e": 2,
    "t": 3,
    "o": 4,
    "l": 5,
    "h": 6,
}


class Solution:
    def Leetcode(self, words: List[str]) -> int:
        @lru_cache(None)
        def dfs(
            index: int, C: int, D: int, E: int, T: int, O: int, L: int, H: int, remain: int
        ) -> int:
            if remain == 0:
                return 0
            if index == n:
                return INF

            res = INF
            word = words[index]
            for state in states[index]:
                count = 0
                counter = [0] * 7
                select = [False] * len(word)
                for i in range(len(word)):
                    if state & (1 << i):
                        count += 1
                        counter[ORDER[word[i]]] += 1
                        select[i] = True
                if (
                    counter[0] > C
                    or counter[1] > D
                    or counter[2] > E
                    or counter[3] > T
                    or counter[4] > O
                    or counter[5] > L
                    or counter[6] > H
                ):
                    continue

                cost = calCost(tuple(select))
                cand = cost + dfs(
                    index + 1,
                    C - counter[0],
                    D - counter[1],
                    E - counter[2],
                    T - counter[3],
                    O - counter[4],
                    L - counter[5],
                    H - counter[6],
                    remain - count,
                )
                if cand < res:
                    res = cand

            return res

        n = len(words)
        states = [[] for _ in range(n)]
        for i, word in enumerate(words):
            m = len(word)
            for state in range(1 << m):
                for j in range(m):
                    if (state >> j) & 1 and word[j] not in CDETOLH:
                        break
                else:
                    states[i].append(state)

        res = dfs(0, 1, 1, 4, 1, 2, 3, 1, 13)
        dfs.cache_clear()
        return res if res != INF else -1


print(Solution().Leetcode(["hello", "leetcode"]))
print(Solution().Leetcode(words=["hold", "engineer", "cost", "level"]))
print(
    Solution().Leetcode(
        words=[
            "lkhqjztn",
            "cpoipalb",
            "hrke",
            "fveuttt",
            "conrzlm",
            "tdrohwgm",
            "odzetred",
            "jekj",
            "lh",
            "kelzwh",
        ]
    )
)

print(Solution().Leetcode(words=["ecleob", "rho", "tw", "lpl", "ebolddec"]))

File: 11_动态规划/经典题/LCP 69. Hello LeetCode!.java
// !java 预处理+多维dp

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

// 11_动态规划/经典题/LCP 69. Hello LeetCode!.py
// 1. 考虑 dp, $dp[index][C][D][E][T][O][L][H]表示 前 index 个单词中,每种字母还要取多少个
// 2. 对每个单词，枚举子集预处理合理的转移状态
// 3. 当删除的字符固定时，从两端向中间删除字符，双指针或 deque 计算最小删除成本 (实际上，这里的成本也可以预处理)
// dp 的状态数为 $n*960$，转移的复杂度为 $O(2^{len(w)}*len(w))$
class Solution {
  private static final String CDETOLH = "cdetolh";

  public int Leetcode(String[] words) {
    int n = words.length;

    // !1.枚举子集预处理合理的转移状态
    List<Integer>[] states = new List[n];
    for (int i = 0; i < n; i++) {
      states[i] = new ArrayList<>();
      int m = words[i].length();
      for (int state = 0; state < 1 << m; state++) {
        boolean isValid = true;
        for (int j = 0; j < m; j++) {
          if (((state >> j) & 1) != 0 && CDETOLH.indexOf(words[i].charAt(j)) == -1) {
            isValid = false;
            break;
          }
        }

        if (isValid) {
          states[i].add(state);
        }
      }
    }

    int[][][][][][][] dp = new int[2][2][5][2][3][4][2], ndp = new int[2][2][5][2][3][4][2]; // CDETOLH
    for (int c = 0; c < 2; ++c) {
      for (int d = 0; d < 2; ++d) {
        for (int e = 0; e < 5; ++e) {
          for (int t = 0; t < 2; ++t) {
            for (int o = 0; o < 3; ++o) {
              for (int l = 0; l < 4; ++l) {
                for (int h = 0; h < 2; ++h) {
                  dp[c][d][e][t][o][l][h] = ndp[c][d][e][t][o][l][h] = -1;
                }
              }
            }
          }
        }
      }
    }
    dp[1][1][4][1][2][3][1] = 0;

    for (int i = 0; i < n; i++) {
      for (int state : states[i]) {
        int[] C = new int[7];
        int m = words[i].length();

        // !2.双指针或deque计算最小删除成本
        Deque<Integer> queue = new ArrayDeque<>();
        for (int j = 0; j < m; j++) {
          if (((state >> j) & 1) != 0) {
            C[CDETOLH.indexOf(words[i].charAt(j))]++;
            queue.offerLast(j);
          }
        }

        int cost = 0;
        int leftMoved = 0, rightMoved = 0;
        while (!queue.isEmpty()) {
          int left = queue.peekFirst(), right = queue.peekLast();
          int leftCost = (left - leftMoved) * (m - 1 - left - rightMoved);
          int rightCost = (right - leftMoved) * (m - 1 - right - rightMoved);
          if (leftCost <= rightCost) {
            queue.pollFirst();
            cost += leftCost;
            leftMoved++;
          } else {
            queue.pollLast();
            cost += rightCost;
            rightMoved++;
          }
        }

        // !3.状态转移
        for (int c = C[0]; c < 2; c++) {
          for (int d = C[1]; d < 2; d++) {
            for (int e = C[2]; e < 5; e++) {
              for (int t = C[3]; t < 2; t++) {
                for (int o = C[4]; o < 3; o++) {
                  for (int l = C[5]; l < 4; l++) {
                    for (int h = C[6]; h < 2; h++) {
                      if (dp[c][d][e][t][o][l][h] != -1) {
                        if (ndp[c - C[0]][d - C[1]][e - C[2]][t - C[3]][o - C[4]][l - C[5]][h - C[6]] == -1) {
                          ndp[c - C[0]][d - C[1]][e - C[2]][t - C[3]][o - C[4]][l - C[5]][h - C[6]] = cost
                              + dp[c][d][e][t][o][l][h];
                        } else {
                          ndp[c - C[0]][d - C[1]][e - C[2]][t - C[3]][o - C[4]][l - C[5]][h - C[6]] = Math.min(
                              ndp[c - C[0]][d - C[1]][e - C[2]][t - C[3]][o - C[4]][l - C[5]][h - C[6]],
                              dp[c][d][e][t][o][l][h] + cost);
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }

      for (int d = 0; d < 2; d++) {
        for (int c = 0; c < 2; c++) {
          for (int e = 0; e < 5; e++) {
            for (int t = 0; t < 2; t++) {
              for (int o = 0; o < 3; o++) {
                for (int l = 0; l < 4; l++) {
                  for (int h = 0; h < 2; h++) {
                    dp[c][d][e][t][o][l][h] = ndp[c][d][e][t][o][l][h];
                  }
                }
              }
            }
          }
        }
      }
    }

    return dp[0][0][0][0][0][0][0];
  }

  public static void main(String[] args) {
    Solution solution = new Solution();
    String[] words = new String[] { "hold", "engineer", "cost", "level" };
    int res = solution.Leetcode(words);
    System.out.println(res);
  }
}